crates/sui-core/src/authority/authority_per_epoch_store_pruner.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::authority::authority_per_epoch_store::EPOCH_DB_PREFIX;
+use std::fs;
+use std::path::PathBuf;
+use std::time::Duration;
+use sui_config::node::AuthorityStorePruningConfig;
+use tokio::sync::oneshot;
+use tracing::log::{error, info};
+use typed_store::rocks::safe_drop_db;
+
+pub struct AuthorityPerEpochStorePruner {
+    _cancel_handle: oneshot::Sender<()>,
+}
+
+impl AuthorityPerEpochStorePruner {
+    pub fn new(parent_path: PathBuf, config: &AuthorityStorePruningConfig) -> Self {
+        let (_cancel_handle, mut recv) = tokio::sync::oneshot::channel();
+        let num_latest_epoch_dbs_to_retain = config.num_latest_epoch_dbs_to_retain;
+        if num_latest_epoch_dbs_to_retain == 0 || num_latest_epoch_dbs_to_retain == u64::MAX {
+            info!("Skipping pruning of epoch tables as we want to retain all versions");
+            return Self { _cancel_handle };
+        }
+        let mut prune_interval =
+            tokio::time::interval(Duration::from_secs(config.epoch_db_pruning_period_secs));
+        tokio::task::spawn(async move {
+            loop {
+                tokio::select! {
+                    _ = prune_interval.tick() => {
+                        info!("Starting pruning of epoch tables");
+                        match Self::prune_old_directories(&parent_path, num_latest_epoch_dbs_to_retain) {
+                            Ok(pruned_count) => info!("Finished pruning old epoch databases. Pruned {} dbs", pruned_count),
+                            Err(err) => error!("Error while removing old epoch databases {:?}", err),
+                        }
+                    }
+                    _ = &mut recv => break,
+                }
+            }
+        });
+        Self { _cancel_handle }
+    }
+
+    fn prune_old_directories(
+        parent_path: &PathBuf,
+        num_latest_epoch_dbs_to_retain: u64,
+    ) -> Result<usize, anyhow::Error> {
+        let mut pruned = 0;
+        let mut candidates = vec![];
+
+        let directories = fs::read_dir(parent_path)?.collect::<Result<Vec<_>, _>>()?;
+        for directory in directories {
+            let path = directory.path();
+            if let Some(filename) = directory.file_name().to_str() {
+                if let Ok(epoch) = filename.split_at(EPOCH_DB_PREFIX.len()).1.parse::<u64>() {
+                    candidates.push((epoch, path));
+                }
+            }
+        }
+        if let Some((latest_epoch, _)) = candidates.iter().max().cloned() {
+            for (epoch, path) in candidates {
+                if epoch + num_latest_epoch_dbs_to_retain <= latest_epoch {
+                    pruned += 1;
+                    safe_drop_db(path.join("recovery_log"))?;
+                    safe_drop_db(path)?;
```

Comment:
Should we check if the "recovery_log" path exists or is it safe to call destroy if for some reason it didn't exist?

Reply:
it's safe to call if it doesn't exist
this API call removes all rocksdb-related files if they exist and directory itself if it's empty afterwards